src/qibo/backends/__init__.py

```diff
@@ -0,0 +1,113 @@
+from qibo import config
+from qibo.config import raise_error, log, warnings, BACKEND_NAME
+from qibo.backends.numpy import NumpyBackend
+from qibo.backends.tensorflow import TensorflowBackend
+
+
+CONSTRUCTED_BACKENDS = {}
+def construct_backend(name):
+    if name not in CONSTRUCTED_BACKENDS:
+        if name == "numpy":
+            CONSTRUCTED_BACKENDS["numpy"] = NumpyBackend()
+        elif name == "tensorflow":
+            CONSTRUCTED_BACKENDS["tensorflow"] = TensorflowBackend()
+        else:
+            raise_error(ValueError, "Unknown backend name {}.".format(name))
+    return CONSTRUCTED_BACKENDS.get(name)
+
+numpy_backend = construct_backend("numpy")
+numpy_matrices = numpy_backend.matrices
+
+if BACKEND_NAME == "tensorflow":
+    K = TensorflowBackend()
+else: # pragma: no cover
+    # CI uses tensorflow as default backend
+    K = NumpyBackend()
```

Comment:
Could you have a backend switcher function between tf and numpy?
We may rename the `set_backend` function below with `set_computational_engine` or `set_computational_model`.

Reply:
To my understanding there are two approaches on how the backend/model can be selected. 

1. What I implemented here, where `set_backend` handles both the backend/engine (numpy or tensorflow) and the model used for gates (custom, defaulteinsum, matmuleinsum). 
2. Have two different switchers, for example `set_backend` for selecting the engine (numpy or tensorflow) and `set_computational_model` (or `set_gate_model`) for selecting the gate model (custom, defaulteinsum or matmuleinsum). 

Personally I am not sure which approach is better. This may be clearer when we implement a third backend which may not support all gate models. An advantage of the first approach was that it was easier to integrate with our current tests which use `set_backend`, however this can be updated as I am revising all tests in #293 anyway.

Let me know which approach you prefer.

Reply:
Thanks for the clarification, we can keep your original implementation (point 1) and then see how it handles different scenarios.